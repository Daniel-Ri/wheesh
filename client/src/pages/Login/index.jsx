import { connect, useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';

// eslint-disable-next-line import/no-absolute-path
import { useEffect, useState } from 'react';
import { Button, FormControl, IconButton, InputAdornment, OutlinedInput } from '@mui/material';
import { Visibility, VisibilityOff } from '@mui/icons-material';
import { useNavigate } from 'react-router-dom';
import { loginUser } from '@containers/Client/actions';
import toast from 'react-hot-toast';
import { createStructuredSelector } from 'reselect';
import { selectLogin } from '@containers/Client/selectors';

import classes from './style.module.scss';

const Login = ({ login, intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [inputs, setInputs] = useState({
    usernameOrEmail: '',
    password: '',
  });
  const [showPassword, setShowPassword] = useState(false);
  const [ableLogin, setAbleLogin] = useState(false);

  const handleClickShowPassword = () => setShowPassword((show) => !show);

  const handleMouseDownPassword = (event) => {
    event.preventDefault();
  };

  const handleInputChange = (e) => {
    setInputs((prev) => ({ ...prev, [e.target.name]: e.target.value }));
  };

  const handleSuccess = () => {
    navigate('/');
  };

  const handleError = (errorMsg) => {
    toast.error(errorMsg);
    setInputs({ usernameOrEmail: '', password: '' });
  };

  const clickLogin = () => {
    dispatch(loginUser(inputs, handleSuccess, handleError));
  };

  useEffect(() => {
    if (login) {
      navigate('/');
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [login]);

  useEffect(() => {
    if (inputs.usernameOrEmail && inputs.password) {
      setAbleLogin(true);
    } else {
      setAbleLogin(false);
    }
  }, [inputs.usernameOrEmail, inputs.password]);

  return (
    <main data-testid="Login" className={classes.main}>
      <div className={classes.container}>
        <header>
          <img src="/wheesh.png" alt="App Icon" />
        </header>
        <div className={classes.message}>
          <div className={classes.mainMsg}>
            <FormattedMessage id="app_please_enter_account_and_password" />
          </div>
          <div className={classes.subMsg}>
            <FormattedMessage id="app_please_log_in_using_registered" />
          </div>
        </div>
        <form>
          <input
            type="text"
            name="usernameOrEmail"
            id="usernameOrEmail"
            value={inputs.usernameOrEmail}
            onChange={handleInputChange}
            placeholder={formatMessage({ id: 'app_enter_username_or_email' })}
            autoComplete="off"
          />
          <FormControl variant="outlined">
            {/* eslint-disable-next-line react/self-closing-comp */}
            {/* <InputLabel htmlFor="outlined-adornment-password"></InputLabel> */}
            <OutlinedInput
              id="outlined-adornment-password"
              name="password"
              placeholder={formatMessage({ id: 'app_enter_the_password' })}
              value={inputs.password}
              onChange={handleInputChange}
              className={classes.input}
              type={showPassword ? 'text' : 'password'}
              endAdornment={
                <InputAdornment position="end">
                  <IconButton
                    aria-label="toggle password visibility"
                    onClick={handleClickShowPassword}
                    onMouseDown={handleMouseDownPassword}
                    edge="end"
                  >
                    {showPassword ? <VisibilityOff /> : <Visibility />}
                  </IconButton>
                </InputAdornment>
              }
              label=""
            />
          </FormControl>
        </form>
        <div className={classes.buttons}>
          <Button variant="contained" disabled={!ableLogin} className={classes.login} onClick={clickLogin}>
            Login
          </Button>
          <Button variant="outlined" className={classes.register} onClick={() => navigate('/register')}>
            Register
          </Button>
        </div>
      </div>
    </main>
  );
};

Login.propTypes = {
  login: PropTypes.bool,
  intl: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  login: selectLogin,
});

export default injectIntl(connect(mapStateToProps)(Login));
