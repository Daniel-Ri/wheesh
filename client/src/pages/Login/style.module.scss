@import '@styles/base';

// Reset
* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  font-family: Poppins;
}

.main {
  flex-grow: 1;
  margin-left: 200px;
  position: relative;
  background: $color-alabaster;
  display: flex;
  justify-content: center;
  align-items: center;
  width: calc(100vw - 200px);

  .container {
    width: 90%;
    max-width: 30rem;
    @include directionY;

    header {
      margin-top: 1rem;
      @include directionX;
      justify-content: center;
      height: 5rem;

      img {
        height: 100%;
        object-fit: contain;
      }
    }

    .message {
      margin-top: 1rem;
      font-size: 100%;

      .subMsg {
        color: $color-cool-gray;
      }
    }

    form {
      margin-top: 1rem;
      width: 100%;
      @include directionY;
      gap: 1rem;

      input {
        padding: 0.25rem 0.5rem;
        border-radius: 5px;
        font-size: 80%;
      }

      .input {
        font-size: 100%;
      }

      input:focus {
        outline: none;
        border-color: $color-picton-blue;
      }
    }

    .buttons {
      margin-top: 1rem;
      @include directionY;
      gap: 1rem;

      .login {
        background-color: $color-picton-blue-dark-shade;
        color: $color-white;
        transition: 0.3s opacity;
      }

      .login:disabled,
      .login[disabled]{
        opacity: 0.5;
      }

      .register {
        color: $color-picton-blue-dark-shade;
      }
    }
  }

  @media screen and (max-width: $screen-lg-limit) {
    margin-left: 192px;
    width: calc(100vw - 192px);
  }

  @media screen and (max-width: $screen-md-limit) {
    margin-left: 10rem;
    width: calc(100vw - 10rem);
  }

  @media screen and (max-width: $screen-sm-limit) {
    margin-left: 0;
    width: 100vw;

    .container {
      header {
        height: 3rem;
      }
    }
  }
}