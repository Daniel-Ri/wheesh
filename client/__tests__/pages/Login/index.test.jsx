import { render } from '@utils/testHelper';
import Login from '@pages/Login';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

const mockProps = {
  login: false,
};

beforeEach(() => {
  wrapper = render(<Login {...mockProps} />);
});

describe('Page Login', () => {
  test('Renders Login page', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('Login')).toBeInTheDocument();
  });

  test('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
