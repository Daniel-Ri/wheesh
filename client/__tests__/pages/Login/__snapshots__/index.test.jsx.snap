// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Login Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="Login"
      >
        <div
          class="container"
        >
          <header>
            <img
              alt="App Icon"
              src="/wheesh.png"
            />
          </header>
          <div
            class="message"
          >
            <div
              class="mainMsg"
            >
              Silakan masukkan akun dan kata sandi Anda
            </div>
            <div
              class="subMsg"
            >
              Silakan login menggunakan nama pengguna atau email Anda yang terdaftar
            </div>
          </div>
          <form>
            <input
              autocomplete="off"
              id="usernameOrEmail"
              name="usernameOrEmail"
              placeholder="Masukkan username atau email"
              type="text"
              value=""
            />
            <div
              class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd input css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  id="outlined-adornment-password"
                  name="password"
                  placeholder="Masukkan password"
                  type="password"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                >
                  <button
                    aria-label="toggle password visibility"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="VisibilityIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </form>
          <div
            class="buttons"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium login css-sghohy-MuiButtonBase-root-MuiButton-root"
              disabled=""
              tabindex="-1"
              type="button"
            >
              Login
            </button>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium register css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Register
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="Login"
    >
      <div
        class="container"
      >
        <header>
          <img
            alt="App Icon"
            src="/wheesh.png"
          />
        </header>
        <div
          class="message"
        >
          <div
            class="mainMsg"
          >
            Silakan masukkan akun dan kata sandi Anda
          </div>
          <div
            class="subMsg"
          >
            Silakan login menggunakan nama pengguna atau email Anda yang terdaftar
          </div>
        </div>
        <form>
          <input
            autocomplete="off"
            id="usernameOrEmail"
            name="usernameOrEmail"
            placeholder="Masukkan username atau email"
            type="text"
            value=""
          />
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd input css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                id="outlined-adornment-password"
                name="password"
                placeholder="Masukkan password"
                type="password"
                value=""
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <button
                  aria-label="toggle password visibility"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="VisibilityIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </form>
        <div
          class="buttons"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium login css-sghohy-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Login
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium register css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Register
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
